### PR TITLE
Fix race condition in MkdirFunc

### PR DESCRIPTION
**Pull request by vit9696**
With variant dirs SCons creates directories in the build directory per each SCons file. And this can race when parallel SCons tasks are launched in the same directory even if they have separate targets and sconsign files.

The issue code in question is as follows:

* `_SConscript` performs directory creation.
* It calls `Dir`'s `_create()` method
* and that one reaches `MkdirFunc` through `MkdirBuilder` (`SCons.Node.Node.build()`)

`MkdirFunc` function already checks for `t.exists()`, so as discussed on the chat a workaround to make this function atomic should be more or less ok.

## Contributor Checklist:

* [ ] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [ ] I have updated the appropriate documentation

No documentation or test changes should be necessary for this commit.

CC @mwichmann

### CODE REVIEW

**Comment by mwichmann** [comment]
Note the one Windows test suite failure is unrelated to the change, it's a different race, which #4402 was just merged to help with - though that came too late for this build.

```
[00:21:32] JSONDecodeError: Expecting value: line 1 column 1 (char 0):
[00:21:32]   File "C:\Users\appveyor\AppData\Local\Temp\1\testcmd.1860.nq3xr9bv\SConstruct", line 3:
[00:21:32]     env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.3', tools=['msvc'])
[00:21:32]   File "C:\projects\scons\SCons\Environment.py", line 1244:
```

File: SCons/Node/FS.py [hunk 1]
```diff
@@ -333,15 +333,24 @@ def UnlinkFunc(target, source, env) -> int:
...
-    # This os.path.exists test looks redundant, but it's possible
-    # when using Install() to install multiple dirs outside the
-    # source tree to get a case where t.exists() is true but
-    # the path does already exist, so this prevents spurious
-    # build failures in that case. See test/Install/multi-dir.
-    if not t.exists() and not os.path.exists(t.get_abspath()):
-        t.fs.mkdir(t.get_abspath())
+    # - It's possible when using Install() to install multiple
+    #   dirs outside the source tree to get a case where t.exists()
+    #   is true but the path does already exist.
```
Comment (mwichmann) [inline SCons/Node/FS.py]: should that say "false"?

Comment (vit9696) [reply SCons/Node/FS.py]: Nice catch!

**Comment by mwichmann** [comment]
As mentioned before this became a PR, this looks simple and straightforward to me.

File: CHANGES.txt [hunk 2]
```diff
@@ -150,6 +150,10 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
     - Fix another instance of `int main()` in CheckLib() causing failures
       when using -Wstrict-prototypes.
 
+  From Vitaly Cheptsov:
+    - Fix race condition in `Mkdir` which can happen when two `SConscript`
+      are parsed simultaneously by two separate build commands.
```
Comment (bdbaddog) [inline CHANGES.txt]: change 'parsed' to 'processed' ?

**Comment by bdbaddog** [comment]
Looks good to me. 
Can you add same blurb from CHANGES.txt to RELEASE.txt?

Also, entries in CHANGES.txt are generally sorted per release by the contributors last name.
Can you move your update to CHANGES.txt to be ordered as such and not at the end of the changes for the current (working) release?

Thanks for this fix!

File: CHANGES.txt (hunk 2 above)
Comment (vit9696) [reply CHANGES.txt]: Sure, changed here and in the code comment as well.

**Comment by vit9696** [comment]
Done. Also rebased and squashed to fix type annotation for the function header.

**Pull request merged**